### ISSUE TITLE
How to incorporate other datasets like Gibson

### ISSUE THREAD

**Issue by gaojunjie1999** [SJTU-ViSYS-team/VisFly#1]
Thanks for your work, could you please inform me where is the habitat rendering engine used? As I don't see "make_sim" function in the code. So I don't know how to use other dataset like Gibson.

**Comment by Fanxing-LI** [comment]
Rendering engine is used in utils.scenedatasets.py. You need to download other datasets in datasets/, and include the folder address of the scenes you wanna use, as "scene_kwargs" and input it to xxxEnv. Maybe you can find an example in **Complete Environment Definition**.

**Comment by gaojunjie1999** [comment]
Thanks, also I wonder if it is possible to train uav to perform tasks like image-goal-navigation/object-goal navigation like in habitat-lab?

**Comment by Fanxing-LI** [comment]
Of course.

You need to define another Env. 
Use RGB camera sensor, use drone local or external estimator(such as, the distance to object) to give appropriate reward, and provide success signal while achieving target (catch object in sight).
That's just my initial and immature thought, you should carefully design and enrich the details.

**Comment by gaojunjie1999** [comment]
Great thanks for your advice

**Comment by gaojunjie1999** [comment]
Correct me if I'm wrong. I noticed that although VecEnvis the base class, it is not used for parallel scene training with multi-thread communication. So I wonder if the parallelization of this env is achieved by training multiple agents in one scene. If that is the case, I wonder if the training speed here is as fast as those using multi-thread parallel scene training architechture?

**Comment by Fanxing-LI** [comment]
It is available to trained multiple agents in one scene. This simulator just inherits VecEnv's interfaces. The bottleneck of speed is rendering the vision. In habitatsim, the graphical computation is deployed on GPUs. I am not sure which one could be faster because I dont know how much CPU resources will be used.

**Comment by gaojunjie1999** [comment]
Thanks for your patient reply. Also, what happens if collisions occur in the simulation, will the drone be bumped back of the drone just slides to a place out of the obstacle? Or will the drone stop?
![屏幕截图 2024-09-29 142459](https://github.com/user-attachments/assets/8569700d-b925-442b-aa60-9fa68f076d62)
As you can see in the snapshot, the drone keeps hitting the obstacle and the trajectory seems to be twisted.

**Comment by Fanxing-LI** [comment]
In a Gym-wrapped Env (requires_grad=False), drone will be reset back to its spawn position with random states (depends on settings) if it crashes or hits on the  wall.
 
we have sucessfully tested the auto-reset function in these datasets wiich are scanned realworld scenes. Maybe you can plot the distance to closest obstacle to show whether it is a bug or the drone just wander near the wall (not crash).

But indeed, this simulator is primarily validated on replicaCAD datasets, because the coordinates/frame of 3D reconstruction datasets (HM3D etc.) are not well uniformed. It is troublesome to specific initial position for each scene.

**Comment by gaojunjie1999** [comment]
Thanks, again could you tell me how the parallel computation is achieved? Since I have not found codes related to multi-process or multi-thread computation.
Although, I noticed that increasing "num_agent_per_scene" can decrease training time.  I wonder if it is because of some of your designed parallel computation mechanism.

**Comment by Fanxing-LI** [comment]
The parallel compuataion is a standrard batch computation in torch, you can find it in dynamics.py.

You can consider it as:
```
scene1:
 agent(1):
...
 agent(num_agent_per_scene):
scene2:
...

scene(num_scene):
```
The states of all agents (num_scene * num_agnet_per_scene)  are package as a batch, and used to simulate physical process.

**Comment by gaojunjie1999** [comment]
Thanks for your illustration. I noticed that you use enu coord frame for calculation, and the translation is calculated by:
I wonder if the frame axis relationship is: y_habitat = z_enu; x_habitat=x_enu; z_habitat=-y_enu? But your calculation suggests otherwise:
![{D6EC538F-EAB6-45A1-90DB-A3513AA093AD}](https://github.com/user-attachments/assets/1be92334-55f0-43e3-82f3-8a2db9c42bca) 
I wonder if my understanding about the two used frames are incorrect.

**Comment by gaojunjie1999** [comment]
As I increases KPv:
![{D98ADC18-5B92-44DB-B54C-8292E15EB37A}](https://github.com/user-attachments/assets/425fded8-73ed-4cec-ade2-67822886851d)

**Comment by gaojunjie1999** [comment]
Thank you for your illustration. However when I use velocity command, and set vx=vy=vz=0. The yaw command leads to great changes of v_z: could you tell me how to fix this?
![{57CFD4DD-FE04-4255-BCFA-5240F56AD898}](https://github.com/user-attachments/assets/99da6627-068d-4112-99d5-6be6563bb39d)

**Comment by gaojunjie1999** [comment]
Please correct me if I am wrong, It seems that you use "return th.stack([roll, pitch, yaw]).flipud()" in toEuler() function, which returns[yaw pitch roll]. But the cotroller calculates v_yaw_des by " yaw_spd_des = (command[0] - self._orientation.toEuler()[2]) * self._VELOCITY_PID.p[2]", where  self._orientation.toEuler()[2] corresponds to the roll angle, not yaw angle.
